### PR TITLE
CB-31 Pause Support for Various Conditions

### PR DESCRIPTION
**Pull request by jmittert**
Added functions for checking for Invisibility. They needed an new ExpectedCondition in order to catch ElementNotFoundExceptions and the like.

### CODE REVIEW

**Comment by Eriten** [comment]
The PR looks correct to me. I assume that you have done some testing; we should write some unit tests to test the invisibility conditions in [`WebDriverElementSystemTest`](https://github.com/sugarcrm/candybean/blob/dev/src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverElementSystemTest.java).

It also might be a good idea to implement waitForInvisible in VoodooGrimoire.

**Comment by jmittert** [comment]
I added a unit test to WebDriverElementSystemTest. I was just going to copy the pauseUntilVisibleTest, but it looks like the website that it depends on now has an anti bot filter that makes the test fail. Do you think it would be a reasonable idea to write a local webpage with the needed testing elements and test from that instead? If not, I can just fix that test in another commit.

I've also implemented waitForInvisible in Voodoo, but there's no use PRing it until Voodoo has access to the waitForInvisible in Candybean.

**Comment by jmittert** [comment]
To complete CB-31, additionally added waitForElement/waitForElementRemoved, waitForAttribute/waitForAttributeRegex, and waitForOnScreen.

File: src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java [hunk 1]
```diff
@@ -1,10 +1,16 @@
 package com.sugarcrm.candybean.automation.webdriver;
 
+import com.gargoylesoftware.htmlunit.ElementNotFoundException;
```
Comment (dsafar) [inline src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java]: I don't see this getting used anywhere.  Is it actually needed?

Comment (jmittert) [inline src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java]: I don't remember adding that, so no.

File: src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java [hunk 2]
```diff
@@ -36,17 +43,16 @@ private static WebDriverElement createWebDriverElement(Hook hook, WebElement ele
...
 	 */
 	private static WebElement findElement(Hook hook, WebDriver driver) throws CandybeanException {
-		try {
-			return driver.findElement(getBy(hook));
-		} catch (WebDriverException e) {
-			throw e;
-		}
+        List<WebElement> elements = driver.findElements(getBy(hook));
+        if (elements.isEmpty())  throw new CandybeanException("No such elements found");
+        return elements.get(0);
```
Comment (dsafar) [inline src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java]: The three lines above should be formatted with tabs, not spaces.

File: src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java [hunk 3]
```diff
@@ -112,10 +116,47 @@ public String toString() {
 	 * Wait until the element is not present on the DOM OR invisible
```
Comment (dsafar) [inline src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java]: Please JavaDoc why not(visible()) isn't sufficient for this.

File: src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java [hunk 4]
```diff
@@ -112,10 +116,47 @@ public String toString() {
...
 	 * @param hook
 	 * @return
-	 * @throws CandybeanException
 	 */
-	public static ExpectedCondition<Boolean> invisible(Hook hook) throws CandybeanException {
-		return ExpectedConditions.invisibilityOfElementLocated(getBy(hook));
+	public static ExpectedCondition<Boolean> invisible(final Hook hook) {
```
Comment (dsafar) [inline src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java]: Please JavaDoc why not(visible()) isn't sufficient for this.

File: src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java [hunk 5]
```diff
@@ -337,4 +410,81 @@ public String toString() {
...
+	 * or the reverse if not.
+	 *
+	 * @param hook	The hook used to find the element
+	 * @param attribute	The attribute to check
+	 * @param value	The expected value of the attribute
+	 * @param expectValue If the value is expected or not
+	 * @return The element if the specified value contains the specified attributed, null otherwise
+	 */
+	public static ExpectedCondition<WebDriverElement> hasAttribute(final Hook hook, final String attribute,
+																   final String value, final boolean expectValue) {
```
Comment (dsafar) [inline src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java]: Formatting?  Way too much indent on this line.

File: src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java [hunk 6]
```diff
@@ -337,4 +410,81 @@ public String toString() {
...
+			public WebDriverElement apply(WebDriver driver) {
+				try {
+					WebElement element = findElement(hook, driver);
+					/*
+					Split the string so that we can match the attribute.
+					If we are waiting for a value of "red", then "red left-aligned large"
+					should match that, but "starred" should not.
+					 */
+					String[] values = element.getAttribute(attribute).split("\\s");
+					for (int x = 0; x < values.length; x++) {
```
Comment (dsafar) [inline src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java]: We generally prefer foreach syntax, e.g. "for(String value : values)" for readability.

File: src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java [hunk 7]
```diff
@@ -337,4 +410,81 @@ public String toString() {
...
+	 *
+	 * @param hook The hook used to find the element
+	 * @param attribute The attribute to check
+	 * @param regex String regex of the expected value of the attribute
+	 * @param expectValue If the value is expected or not
+	 * @return The element if the specified value contains the specified attributed, null otherwise
+	 * @throws CandybeanException If the element is not found
+	 */
+	public static ExpectedCondition<WebDriverElement> hasRegexAttribute(final Hook hook, final String attribute,
+																		final String regex, final boolean expectValue)
```
Comment (dsafar) [inline src/main/java/com/sugarcrm/candybean/automation/webdriver/WaitConditions.java]: Too much indent again.

**Comment by jmittert** [comment]
Fixed whitespace and documentation issues.

File: src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverElementSystemTest.java [hunk 8]
```diff
@@ -343,6 +374,58 @@ public void getCssValueTest() throws CandybeanException {
...
 
 	@Test
+	public void waitForAttribute() throws Exception {
+		iface.go("file://"+ System.getProperty("user.dir")+"/resources/html/test/blinking.html");
+		WebDriverElement text = iface.getWebDriverElement(new Hook(Strategy.ID, "p1"));
+		text.click();
+		Assert.assertTrue(iface.getPause().waitForAttribute(new Hook(Strategy.ID, "p1"), "class", "hidden", true, 10) != null);
+		Assert.assertTrue(iface.getPause().waitForAttribute(new Hook(Strategy.ID, "p1"), "class", "hidden", false, 10) != null);
+		Assert.assertTrue(iface.getPause().waitForAttribute(new Hook(Strategy.ID, "p1"), "class", "normal", false, 10) != null);
+		Assert.assertTrue(iface.getPause().waitForAttribute(new Hook(Strategy.ID, "p1"), "class", "normal", false, 10) != null);
```
Comment (dsafar) [inline src/test/java/com/sugarcrm/candybean/automation/webdriver/WebDriverElementSystemTest.java]: Two lines in a row look identical -- is this intentional, or should one be passing true?

**Comment by dsafar** [comment]
Ready for updates and needs a rebase so it can be merged.

**Comment by jmittert** [comment]
Updated and rebased.

**Pull request merged**